File: ext/post-to-tumblr.js
import fs from "fs";
import tumblr from "tumblr.js";

/* =====================
 * Konfigurasi
 * ===================== */
const ARTICLE_FILE = "artikel.json";
const DATABASE_FILE = "mini/posted-tumblr.txt";
const BLOG_NAME = "frijal";
const BASE_URL = "https://dalam.web.id";

const client = tumblr.createClient({
  consumer_key: process.env.TUMBLR_CONSUMER_KEY,
  consumer_secret: process.env.TUMBLR_CONSUMER_SECRET,
  token: process.env.TUMBLR_TOKEN,
  token_secret: process.env.TUMBLR_TOKEN_SECRET,
});

/* =====================
 * Util
 * ===================== */
const cleanTag = (str) =>
str.replace(/&/g, "dan").replace(/[^\w\s]/g, "").trim();

const slugify = (text) =>
text.toLowerCase().trim().replace(/\s+/g, '-');

/* =====================
 * Load Database
 * ===================== */
let postedDatabase = "";
if (fs.existsSync(DATABASE_FILE)) {
  postedDatabase = fs.readFileSync(DATABASE_FILE, "utf8");
}

/* =====================
 * Load & Cari Artikel
 * ===================== */
const raw = JSON.parse(fs.readFileSync(ARTICLE_FILE, "utf8"));
let allArticles = [];

for (const [category, items] of Object.entries(raw)) {
  const catSlug = slugify(category); // <--- Pakai catSlug

  for (const item of items) {
    const fileSlug = item[1].replace('.html', '').replace(/^\//, '');

    // Perbaikan di sini: Pastikan variabelnya sama (catSlug)
    const fullUrl = `${BASE_URL}/${catSlug}/${fileSlug}`;

    // Cek slug di database biar nggak dobel post
    const isPosted = postedDatabase.includes(fileSlug);

    if (!isPosted) {
      allArticles.push({
        title: item[0],
        url: fullUrl,
        slug: fileSlug,
        date: item[3],
        desc: item[4] || "Archive.",
        category: category
      });
    }
  }
}

// Sorting terbaru
allArticles.sort((a, b) => b.date.localeCompare(a.date));

const target = allArticles[0];

if (!target) {
  console.log("✅ Tumblr: Tidak ada artikel baru (berdasarkan slug).");
  process.exit(0);
}

/* =====================
 * Posting
 * ===================== */
const tags = ["fediverse", "Repost", "Ngopi", "Indonesia", target.category];

console.log(`🚀 Mengirim ke Tumblr: ${target.title}`);

const content = [
  { type: "text", text: target.desc },
{ type: "text", text: "#fediverse #Indonesia" },
{ type: "link", url: target.url }
];

client.createPost(BLOG_NAME, {
  content: content,
  tags: tags.map(t => cleanTag(t))
}, (err, data) => {
  if (err) {
    console.error("❌ Gagal post Tumblr:", JSON.stringify(err, null, 2));
    process.exit(1);
  }

  if (!fs.existsSync("mini")) fs.mkdirSync("mini", { recursive: true });
  fs.appendFileSync(DATABASE_FILE, target.url + "\n");

  console.log("✅ Berhasil post ke Tumblr:", target.url);
  setTimeout(() => process.exit(0), 500);
});

File: alat/post-to-tumblr.rs
use serde::Deserialize;
use serde_json::json;
use std::collections::HashMap;
use std::env;
use std::fs;
use std::path::Path;
use url::Url;

/**
 * CONFIGURATION
 */
const JSON_FILE: &str = "artikel.json";
const DATABASE_FILE: &str = "mini/posted-tumblr.txt";
const BLOG_NAME: &str = "frijal";
const BASE_URL: &str = "https://dalam.web.id";

#[allow(dead_code)]
#[derive(Debug, Deserialize)]
struct Post(String, String, String, String, Option<String>);

fn slugify(text: &str) -> String {
    text.to_lowercase()
        .trim()
        .replace(|c: char| !c.is_alphanumeric() && c != ' ', "")
        .replace(' ', "-")
}

fn clean_tag(t: &str) -> String {
    t.replace('&', "dan")
        .chars()
        .filter(|c| c.is_alphanumeric() || c.is_whitespace())
        .collect::<String>()
        .trim()
        .to_string()
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("🎨 Memulai Tumblr Autopost (Rust Engine)...");

    // 1. Ambil Kunci Rahasia
    let consumer_key = env::var("TUMBLR_CONSUMER_KEY").expect("TUMBLR_CONSUMER_KEY not set");
    let consumer_secret = env::var("TUMBLR_CONSUMER_SECRET").expect("TUMBLR_CONSUMER_SECRET not set");
    let token = env::var("TUMBLR_TOKEN").expect("TUMBLR_TOKEN not set");
    let token_secret = env::var("TUMBLR_TOKEN_SECRET").expect("TUMBLR_TOKEN_SECRET not set");

    // 2. Baca Database
    let posted_database = if Path::new(DATABASE_FILE).exists() {
        fs::read_to_string(DATABASE_FILE)?
    } else {
        String::new()
    };

    // 3. Parse Artikel
    let data_str = fs::read_to_string(JSON_FILE).expect("artikel.json tidak ditemukan");
    let data: HashMap<String, Vec<Post>> = serde_json::from_str(&data_str)?;

    let mut all_articles = Vec::new();
    for (category, posts) in data {
        let cat_slug = slugify(&category);
        for p in posts {
            let file_slug = p.1.trim_start_matches('/').replace(".html", "");
            if file_slug.startsWith("agregat-20") { continue; }

            let full_url = format!("{}/{}/{}", BASE_URL, cat_slug, file_slug);
            if !posted_database.contains(&file_slug) {
                all_articles.push((p, full_url, category.clone()));
            }
        }
    }

    all_articles.sort_by(|a, b| b.0.3.cmp(&a.0.3));

    if all_articles.is_empty() {
        println!("✅ Tumblr: Tidak ada artikel baru.");
        return Ok(());
    }

    let (target_data, target_url, category_name) = &all_articles[0];
    println!("🚀 Mengirim ke Tumblr: {}", target_data.0);

    // 4. Siapkan Konten NPF (Neue Post Format) Tumblr
    let content = json!([
        { "type": "text", "text": target_data.4.as_deref().unwrap_or("Archive.") },
        { "type": "text", "text": "#fediverse #Indonesia" },
        { "type": "link", "url": target_url }
    ]);

    let tags = vec!["fediverse", "Repost", "Ngopi", "Indonesia", category_name];
    let cleaned_tags: String = tags.iter().map(|t| clean_tag(t)).collect::<Vec<_>>().join(",");

    // 5. OAuth Signature & Request
    let api_url = format!("https://api.tumblr.com/v2/blog/{}/posts", BLOG_NAME);
    let consumer = oauth1::Token::new(consumer_key, consumer_secret);
    let access = oauth1::Token::new(token, token_secret);

    let mut params = HashMap::new();
    params.insert("content", content.to_string());
    params.insert("tags", cleaned_tags);

    let header = oauth1::authorize("POST", &api_url, &consumer, Some(&access), Some(params));

    let client = reqwest::Client::new();
    let response = client
        .post(&api_url)
        .header(reqwest::header::AUTHORIZATION, header)
        .form(&[
            ("content", content.to_string()),
            ("tags", cleaned_tags)
        ])
        .send()
        .await?;

    if response.status().is_success() {
        let mut file = fs::OpenOptions::new().append(true).create(true).open(DATABASE_FILE)?;
        use std::io::Write;
        writeln!(file, "{}", target_url)?;
        println!("✅ Berhasil post ke Tumblr!");
    } else {
        eprintln!("❌ Gagal: {}", response.text().await?);
    }

    Ok(())
}
